### PR TITLE
[SPARK-38170][TESTS][SQL] Make ThriftServerWithSparkContextInHttpSuite pass with ANSI enabled

### PR DESCRIPTION
**Pull request by amaliujia**
### What changes were proposed in this pull request?

Fix ThriftServerWithSparkContextInHttpSuite when ANSI mode is on.

### Why are the changes needed?

To help enable SQL `ANSI` mode by default.

### Does this PR introduce _any_ user-facing change?

No. This is testing only change.

### How was this patch tested?

UT

### CODE REVIEW

**Comment by amaliujia** [comment]
R: @gengliangwang

File: sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/ThriftServerWithSparkContextSuite.scala [hunk 1]
```diff
@@ -61,22 +61,32 @@ trait ThriftServerWithSparkContextSuite extends SharedThriftServer {
...
       assert(!e.getMessage.contains("" +
         "java.lang.NumberFormatException: invalid input syntax for type numeric: 1.2"))
       assert(e.getSQLState == "22023")
+
+      val e1 = intercept[HiveSQLException] {
+        exec(s"set ${SQLConf.ANSI_ENABLED.key}=true")
+        exec(sql)
+      }
+
+      assert(e.getMessage
```
Comment (gengliangwang) [inline sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/ThriftServerWithSparkContextSuite.scala:82]: Shall we check e1 instead?

**Comment by gengliangwang** [comment]
@amaliujia We will have one new jobs for ANSI mode on. So, I think we need to read the conf value of ANSI mode and run different checks, instead of running both scenarios in one test case.

**Comment by amaliujia** [comment]
@gengliangwang is there an example of how to read the conf value? Or should I wait until the new job start then continue fixing this test?

**Comment by gengliangwang** [comment]
BTW, how many test suites failed under hive-thriftserver?

**Comment by amaliujia** [comment]
@gengliangwang around 6. 2 are being fixed.

**Comment by amaliujia** [comment]
@gengliangwang 

I updated the test case to address your comments

File: sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/ThriftServerWithSparkContextSuite.scala [hunk 2]
```diff
@@ -57,34 +57,50 @@ trait ThriftServerWithSparkContextSuite extends SharedThriftServer {
...
-      assert(e.getSQLState == "22023")
+
+      if (ansi) {
+        assert(e.getMessage.contains(
+          "java.lang.NumberFormatException: invalid input syntax for type numeric: 1.2"))
+        assert(e.getSQLState == "42000")
+      } else {
+        assert(e.getMessage
+          .contains("The second argument of 'date_sub' function needs to be an integer."))
+        assert(e.getMessage.contains("[SECOND_FUNCTION_ARGUMENT_NOT_INTEGER]"))
```
Comment (allisonwang-db) [inline sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/ThriftServerWithSparkContextSuite.scala:99]: Maybe `e.getErrorClass`?

File: sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/ThriftServerWithSparkContextSuite.scala [hunk 3]
```diff
@@ -57,34 +57,50 @@ trait ThriftServerWithSparkContextSuite extends SharedThriftServer {
...
+        exec(sql)
       }
 
-      assert(e.getMessage
-        .contains("The second argument of 'date_sub' function needs to be an integer."))
-      assert(!e.getMessage.contains("" +
-        "java.lang.NumberFormatException: invalid input syntax for type numeric: 1.2"))
-      assert(e.getSQLState == "22023")
+      if (ansi) {
+        assert(!e.getMessage
```
Comment (allisonwang-db) [inline sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/ThriftServerWithSparkContextSuite.scala:73]: Is the error message change when ANSI is enabled intended?

**Comment by allisonwang-db** [comment]
> @amaliujia We will have one new jobs for ANSI mode on. So, I think we need to read the conf value of ANSI mode and run different checks, instead of running both scenarios in one test case.

@gengliangwang I think a unit test should test both ANSI on and ANSI off behaviors regardless of whether we have a new CI job or not, unless the unit test is testing against a particular ANSI behavior.

File: sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/ThriftServerWithSparkContextSuite.scala (hunk 2 above)
Comment (HyukjinKwon) [reply sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/ThriftServerWithSparkContextSuite.scala:99]: Yeah, I think we can fix together while we're here.

File: sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/ThriftServerWithSparkContextSuite.scala [hunk 4]
```diff
@@ -57,34 +57,50 @@ trait ThriftServerWithSparkContextSuite extends SharedThriftServer {
 
   test("Full stack traces as error message for jdbc or thrift client") {
     val sql = "select date_sub(date'2011-11-11', '1.2')"
+    val ansi: Boolean = SQLConf.get.getConf(SQLConf.ANSI_ENABLED)
+
     withCLIServiceClient() { client =>
       val sessionHandle = client.openSession(user, "")
 
       val confOverlay = new java.util.HashMap[java.lang.String, java.lang.String]
+      val exec: String => OperationHandle = client.executeStatement(sessionHandle, _, confOverlay)
```
Comment (HyukjinKwon) [inline sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/ThriftServerWithSparkContextSuite.scala:66]: Is this change related? I think seems unnecessary

**Comment by amaliujia** [comment]
> > @amaliujia We will have one new jobs for ANSI mode on. So, I think we need to read the conf value of ANSI mode and run different checks, instead of running both scenarios in one test case.
> 
> @gengliangwang I think a unit test should test both ANSI on and ANSI off behaviors regardless of whether we have a new CI job or not, unless the unit test is testing against a particular ANSI behavior.

@gengliangwang please let me know your opinion. I can make appropriate change if we have a consensus here.

**Comment by gengliangwang** [comment]
>> @gengliangwang I think a unit test should test both ANSI on and ANSI off behaviors regardless of whether we have a new CI job or not, unless the unit test is testing against a particular ANSI behavior.

> @gengliangwang please let me know your opinion. I can make appropriate change if we have a consensus here.

I see.  Either way should be fine then

File: sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/ThriftServerWithSparkContextSuite.scala (hunk 2 above)
Comment (amaliujia) [reply sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/ThriftServerWithSparkContextSuite.scala:99]: I think because `e` here is `java.sql.SQLException`, it does not `getErrorClass`. Maybe that's why this code was written as what we see it now.

File: sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/ThriftServerWithSparkContextSuite.scala (hunk 3 above)
Comment (amaliujia) [reply sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/ThriftServerWithSparkContextSuite.scala]: @allisonwang-db as this PR is just fix test case. If we have concern on the error, we can log a JIRA to further investigate but not coupled with this task. What do you think?

File: sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/ThriftServerWithSparkContextSuite.scala (hunk 4 above)
Comment (HyukjinKwon) [reply sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/ThriftServerWithSparkContextSuite.scala:66]: Ah, okay. It was to deduplicate. I think it's fine to don't do this as it's one liner, and making it partial function is less readable IMO. But I am fine.

**Comment by gengliangwang** [comment]
@amaliujia I just tried running the failed test.
I think it would be better to change to query instead, e.g
```
select abs(date'2021-01-1')
```
This query throws the same exception with/without ANSI mode on.

**Comment by gengliangwang** [comment]
Also, it would be better to combine the fix of one or two more test suites together in one PR.

**Comment by amaliujia** [comment]
> @amaliujia I just tried running the failed test. I think it would be better to change to query instead, e.g
> 
> ```
> select abs(date'2021-01-1')
> ```
> 
> This query throws the same exception with/without ANSI mode on.

I see. Interesting. So the goal is to avoid issues brought by enabling ANSI mode on such non-core tests (thus changing query is ok). If so then the change will be easier to make.


> Also, it would be better to combine the fix of one or two more test suites together in one PR.
If we can change queries on Thrift server related tests, then combing multiple fixes into one PR wouldn't too bad.

**Comment by gengliangwang** [comment]
I am fixing this one by using another SQL query: https://github.com/apache/spark/pull/35802

**Comment by gengliangwang** [comment]
I am fixing this one by using another SQL query: https://github.com/apache/spark/pull/35802
